src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Utf8JsonWriterTests.Values.StringSegment.cs

```diff
@@ -0,0 +1,1109 @@
+            jsonUtf8.WriteStringValueSegment("\uD800".AsSpan(), false);
+            jsonUtf8.Flush();
+
+            Assert.Equal(0, jsonUtf8.BytesPending);
+            Assert.Equal(1, jsonUtf8.BytesCommitted);
+
+            jsonUtf8.Reset();
+
+            Assert.Equal(0, jsonUtf8.BytesPending);
+            Assert.Equal(0, jsonUtf8.BytesCommitted);
+
+            jsonUtf8.WriteStringValueSegment("\uDC00".AsSpan(), true);
+
+            string expected = @"""\uFFFD""";
+            Assert.Equal(expected.Length, jsonUtf8.BytesPending);
+            Assert.Equal(0, jsonUtf8.BytesCommitted);
+
+            jsonUtf8.Flush();
+
+            Assert.Equal(0, jsonUtf8.BytesPending);
+            Assert.Equal(expected.Length, jsonUtf8.BytesCommitted);
+            JsonTestHelper.AssertContents('"' + expected, output);
+        }
+
+        [Fact]
+        public static void WriteStringValueSegment_Utf8_Reset()
+        {
+            var output = new ArrayBufferWriter<byte>();
+            using var jsonUtf8 = new Utf8JsonWriter(output);
+
+            jsonUtf8.WriteStringValueSegment([0b110_11111], false);
+            jsonUtf8.Flush();
+
+            Assert.Equal(0, jsonUtf8.BytesPending);
+            Assert.Equal(1, jsonUtf8.BytesCommitted);
+
+            jsonUtf8.Reset();
+
+            Assert.Equal(0, jsonUtf8.BytesPending);
+            Assert.Equal(0, jsonUtf8.BytesCommitted);
+
+            jsonUtf8.WriteStringValueSegment([0b10_111111], true);
+
+            string expected = @"""\uFFFD""";
+            Assert.Equal(expected.Length, jsonUtf8.BytesPending);
+            Assert.Equal(0, jsonUtf8.BytesCommitted);
+
+            jsonUtf8.Flush();
+
+            Assert.Equal(0, jsonUtf8.BytesPending);
+            Assert.Equal(expected.Length, jsonUtf8.BytesCommitted);
+            JsonTestHelper.AssertContents('"' + expected, output);
+        }
+
+        [Fact]
+        public static void WriteStringValueSegment_MixEncoding()
+        {
+            {
+                var output = new ArrayBufferWriter<byte>();
+                using var jsonUtf8 = new Utf8JsonWriter(output);
+
+                // Becuase the first code point is a surrogate pair, it will be
+                // saved until the next write to complete it. It is saved in the
+                // original encoding, UTF-16, so it will be 0b1101_1000 0b1101_1000
+                jsonUtf8.WriteStringValueSegment("\uD8D8".AsSpan(), false);
+
+                // Now we write a UTF-8 continuation byte. With the previous partial
+                // state, the whole sequence is 0b110_11000 0b110_11000 0b10_111111.
+                jsonUtf8.WriteStringValueSegment([0b10_111111], true);
+
+                jsonUtf8.Flush();
+
+                // If this is interpreted as UTF-8, the first byte is invalid because
+                // it is a 2-byte start unit but the second byte is not a continuation.
+                // So a replacement character gets written for the first byte. The second and
```

Comment:
I was meaning to ask, in non-segmented writes do invalid UTF-8 sequences also get written out using the replacement character?

Reply:
Yes, if callers don't mix encodings, the segmented writes will produce the exacts same results as the non-segmented writes (there's no way to mix encodings in non-segmented writes) regardless of the positions at which the string is segmented. A couple of other tests make sure to validate that.